### PR TITLE
SERVER-39191 Optimize count queries on sharded collections

### PR DESCRIPTION
**Pull request by Sebmaster**
When counts were made orphan-aware the addition of the SHARD_FILTER
stage started preventing the COUNT_SCAN optimization on all sharded
collections, leading to ~5x performance degradation due to the necessity
of IXSCAN (and possibly FETCH).

However, there can not be any orphans if the shard key is fully
specified: A mongodb instance will never have orphans for chunk ranges
it owns and mongos will send a query with a full shard key to one
replicaset only.

This allows stripping the SHARDING_FILTER stage for cases where the
query specifies the full shard key, regaining the lost performance.

### CODE REVIEW

**Comment by puppyofkosh** [comment]
I will take some time to review this when my current project wraps up in a few weeks.

File: jstests/sharding/count_queryplan.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+createCollection(coll, {a: 1, b: 1});
+
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true}));
+assertShardFilter(explain);
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true, b: true}));
+assertCountScan(explain);
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true, b: {$in: [true, false]}}));
+assertShardFilter(explain);
+
+
```
Comment (puppyofkosh) [inline jstests/sharding/count_queryplan.js:64]: We avoid multiple lines of vertical whitespace, so delete one of these blank lines.

File: src/mongo/db/query/get_executor.cpp [hunk 2]
```diff
@@ -306,7 +306,20 @@ void fillOutPlannerParams(OperationContext* opCtx,
         auto collDesc = CollectionShardingState::get(opCtx, canonicalQuery->nss())
                             ->getCollectionDescription(opCtx);
         if (collDesc.isSharded()) {
-            plannerParams->shardKey = collDesc.getKeyPattern();
+            const auto keyPattern = collDesc.getKeyPattern();
```
Comment (puppyofkosh) [inline src/mongo/db/query/get_executor.cpp:309]: Suggest using `const auto&` here to avoid copying the key pattern object unnecessarily.

File: src/mongo/db/query/get_executor.cpp [hunk 3]
```diff
@@ -306,7 +306,20 @@ void fillOutPlannerParams(OperationContext* opCtx,
...
+            ShardKeyPattern shardKeyPattern(keyPattern);
+
+            // If we're specifying one shard exactly, we're always going to land on exactly one
```
Comment (puppyofkosh) [inline src/mongo/db/query/get_executor.cpp:312]: I find the first sentence of this comment a little awkward. I think the intention was to say "if we're specifying an exact value for the shard key, we're always going ...". I would suggest rewording the comment to something like this:

```
// If the shard key is specified exactly, the query is guaranteed to only target one
// shard. Shards cannot own orphans for the key ranges they own, so there is no need
// to include a shard filtering stage. By omitting the shard filter, it may be possible
// to get a more efficient plan (for example, a COUNT_SCAN may be used if the query is
// eligible).
```

File: jstests/sharding/count_queryplan.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+jsTest.log('Tests with compound shard key');
+createCollection(coll, {a: 1, b: 1});
+
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true}));
+assertShardFilter(explain);
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true, b: true}));
+assertCountScan(explain);
+explain = assert.commandWorked(coll.explain('executionStats').count({a: true, b: {$in: [true, false]}}));
```
Comment (puppyofkosh) [inline jstests/sharding/count_queryplan.js:61]: This line is just over our 100-char limit, so suggest changing to:
```
explain =
    assert.commandWorked(coll.explain('executionStats').count({a: true, b: {$in: [true, false]}}));
```

**Review by puppyofkosh — CHANGES_REQUESTED** [review]
Hello Sebastian, thank you for your pull request. I reviewed the changes and they look great! All of my comments are relatively minor nitpicks. Most of them are just around our coding style.

If you like, I can deal with the comments and then merge the patch under your name. If you prefer to make the changes yourself that's okay too.

Also, you will have to sign the [Contributor's Agreement](https://www.mongodb.com/legal/contributor-agreement) as part of our considering the PR. 

IB

File: jstests/sharding/count_queryplan.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+assertCountScan(explain);
+
+// We're not checking shard key for equality, therefore need a sharding filter.
+explain = assert.commandWorked(coll.explain('executionStats').count({a: {$in: [true, false]}}));
+assertShardFilter(explain);
+
+// We're requesting a disjoint key from shardkey, therefore need a sharding filter.
+explain = assert.commandWorked(coll.explain('executionStats').count({b: true}));
+assertShardFilter(explain);
+
```
[the remark this answers is not in the document]
Comment (Sebmaster) [reply jstests/sharding/count_queryplan.js:52]: Shall we use `assertNoShardFilter` for the count queries as well then, or keep checking for COUNT_SCAN exactly?

**Comment by Sebmaster** [comment]
> Also, you will have to sign the Contributor's Agreement as part of our considering the PR.

I signed the CLA already, at least I have the confirmation email around. Is it tied to the commit email or GH username, because only the username matches.

I'm working on the style changes, I think I may have used an older file as a template. Might be good to add an example for a file that matches the latest expectations to the wiki/contributor guidelines 😅

**Comment by Sebmaster** [comment]
Turns out the finds are producing query plans with FETCH, so I made the SHARDING_FILTER check accept both IXSCAN and FETCH.

Also took me a while to get the mongo test setup back, but these changes are passing for me again.

File: jstests/sharding/count_queryplan.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (puppyofkosh) [reply jstests/sharding/count_queryplan.js]: What you did looks good.

**Comment by puppyofkosh** [comment]
> I signed the CLA already, at least I have the confirmation email around. Is it tied to the commit email or GH username? , Only the username matches in this case.

Ok, got it. I think just the github username is what matters.

> I'm working on the style changes, I think I may have used an older file as a template. Might be good to add an example for a file that matches the latest expectations to the wiki/contributor guidelines
Yeah, I'll look into this. The JS tests are a real mixed bag.

Anyway, with your latest changes this LGTM! Once I get confirmation you signed the contributor agreement (waiting to hear back from someone) and once https://jira.mongodb.org/browse/SERVER-49802 is done, I'll merge this and close the PR.

**Comment by Sebmaster** [comment]
Nice! Thanks for the good review. I squashed the commits together to make it slightly prettier/easier to merge.

**Comment by puppyofkosh** [comment]
Hi Sebastian, it seems like there's some trouble finding your name in the list of contributors. Do you know which email you used to sign originally? Alternatively, could you re-sign the agreement, probably with the email tied to the commit (smayr [at] atlassian).

**Comment by Sebmaster** [comment]
Original email should've been `mongodb [at] smayr.name`.

**Comment by puppyofkosh** [comment]
Thanks, they were able to find your name. Now we just have to wait for SERVER-50849 to go in.

**Comment by puppyofkosh** [comment]
Alright, this made it through the commit queue. Congrats @Sebmaster and thank you for the contribution.
